compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala

```diff
@@ -2815,9 +2815,179 @@ class JSCodeGen()(using genCtx: Context) {
+              fun1.symbol == jsdefn.Selectable_reflectiveSelectableFromLangReflectiveCalls =>
+        genExpr(receiver)
+
+      case _ =>
+        report.error(
+            "The receiver of Selectable.selectDynamic or Selectable.applyDynamic " +
+            "must be a call to the (implicit) method scala.reflect.Selectable.reflectiveSelectable. " +
+            "Other uses are not supported in Scala.js.",
+            tree.sourcePos)
+        js.Undefined()
+    }
+
+    // Extract the method name as a String
+    val methodNameStr = args.head match {
+      case Literal(Constants.Constant(name: String)) =>
+        name
+      case _ =>
+        report.error(
+            "The method name given to Selectable.selectDynamic or Selectable.applyDynamic " +
+            "must be a literal string. " +
+            "Other uses are not supported in Scala.js.",
+            args.head.sourcePos)
+        "erroneous"
+    }
+
+    val (formalParamTypeRefs, actualArgs) = if (isSelectDynamic) {
+      (Nil, Nil)
+    } else {
+      // Extract the param type refs and actual args from the 2nd and 3rd argument to applyDynamic
+      args.tail match {
+        case WrapArray(classTagsArray: JavaSeqLiteral) :: WrapArray(actualArgsAnyArray: JavaSeqLiteral) :: Nil =>
+          // Extract jstpe.Type's and jstpe.TypeRef's from the ClassTag.apply(_) trees
+          val formalParamTypesAndTypeRefs = classTagsArray.elems.map {
+            case Apply(fun, Literal(const) :: Nil)
+                if fun.symbol == defn.ClassTagModule_apply && const.tag == Constants.ClazzTag =>
+              toIRTypeAndTypeRef(const.typeValue)
+            case classTag =>
+              report.error(
+                  "The ClassTags passed to Selectable.applyDynamic must be " +
+                  "literal ClassTag(classOf[T]) expressions " +
+                  "(typically compiler-generated). " +
+                  "Other uses are not supported in Scala.js.",
+                  classTag.sourcePos)
+              (jstpe.AnyType, jstpe.ClassRef(jsNames.ObjectClass))
+          }
+
+          // Gen the actual args, downcasting them to the formal param types
+          val actualArgs = actualArgsAnyArray.elems.zip(formalParamTypesAndTypeRefs).map {
+            (actualArgAny, formalParamTypeAndTypeRef) =>
+              val genActualArgAny = genExpr(actualArgAny)
+              js.AsInstanceOf(genActualArgAny, formalParamTypeAndTypeRef._1)(genActualArgAny.pos)
+          }
+
+          (formalParamTypesAndTypeRefs.map(_._2), actualArgs)
+
+        case _ =>
+          report.error(
+              "Passing the varargs of Selectable.applyDynamic with `: _*` " +
+              "is not supported in Scala.js.",
+              tree.sourcePos)
+          (Nil, Nil)
+      }
+    }
+
+    val methodName = MethodName.reflectiveProxy(methodNameStr, formalParamTypeRefs)
+
+    js.Apply(js.ApplyFlags.empty, receiver, js.MethodIdent(methodName), actualArgs)(jstpe.AnyType)
```

Comment:
Judging from the code of `reflect.Selectable`, `selectDynamic` is able to select actual bytecode fields (as opposed to accessors). Then again, it falls back to `applyDynamic`.

I do not know the details of how dotty translates fields, but it seems this is problematic for exact compatibility.

Reply:
On the JVM, we need to look up fields because of Java interop. A Java class such as
```java
public class A {
  public int foo = 5;
}
```
qualifies for the Scala structural type
```scala
{ val foo: Int }
```
so for Java we need to try fields.

In the SJSIR, however, since JDK classes are implemented in Scala anyway, I don't think there can be an issue.

Reply:
Ah, OK. Fair enough.